demo-app/src/pages/platform-devices/DevicesMap/DevicesMap.jsx

```diff
@@ -0,0 +1,85 @@
+import React, { useContext } from 'react';
+import { Controls, Map, Marker, MarkerCluster } from 'grommet-leaflet';
+import {
+  Anchor,
+  Box,
+  DataContext,
+  NameValueList,
+  NameValuePair,
+  Text,
+} from 'grommet';
+
+const ClusterDetail = ({ cluster }) => {
+  const childMarkers = cluster.getAllChildMarkers();
+  const locations = {};
+  const deviceTypes = {};
+  const parts = {};
+  childMarkers.forEach(marker => {
+    if (locations[marker.options.detail.properties.location]) {
+      locations[marker.options.detail.properties.location] += 1;
+    } else {
+      locations[marker.options.detail.properties.location] = 1;
+    }
+    if (deviceTypes[marker.options.detail.properties.device_type]) {
+      deviceTypes[marker.options.detail.properties.device_type] += 1;
+    } else {
+      deviceTypes[marker.options.detail.properties.device_type] = 1;
+    }
+    if (parts[marker.options.detail.properties.part]) {
+      parts[marker.options.detail.properties.part] += 1;
+    } else {
+      parts[marker.options.detail.properties.part] = 1;
+    }
+  });
+
+  return (
+    <Box gap="small">
+      <Box gap="xsmall">
+        <Text>{childMarkers.length} devices</Text>
+        <NameValueList nameProps={{ width: 'xsmall' }}>
+          <NameValuePair name="Locations">
+            {Object.entries(locations).length}
+          </NameValuePair>
+          <NameValuePair name="Device types">
+            {Object.entries(deviceTypes).length}
+          </NameValuePair>
+          <NameValuePair name="Parts">
+            {Object.entries(parts).length}
+          </NameValuePair>
+        </NameValueList>
+      </Box>
+      <Anchor alignSelf="end" label="View details" />
+    </Box>
+  );
+};
+
+export const DevicesMap = () => {
+  const mapContainerRef = React.useRef();
+  const dataContext = useContext(DataContext);
+  const { data } = dataContext;
+  const locations = data.map(device => {
+    return device.geometry.coordinates;
+  });
+
+  return (
+    <Map id="map" ref={mapContainerRef}>
+      {data.length >= 1 ? (
+        <>
+          <Controls locations={locations} />
+          <MarkerCluster popup={cluster => <ClusterDetail cluster={cluster} />}>
+            {data.map(device => {
+              return (
+                <Marker
+                  key={device.properties.serial_number}
+                  position={device.geometry.coordinates}
+                  // popup={device.properties.serial_number}
+                  detail={device}
```

Comment:
What is `detail`?

Reply:
We need some sort of mechanism for the object's data to be available to a Marker for popup purposes.

Probably will change this property to `data`.

Reply:
I would envision something like this for the popup:

```
{devices.map(device =>
<Marker
   popup={
      <Box>{device.properties.serial_number}</Box>
   }
/>
}
```
So the caller controls the popup externally and should have access. Will look back at the code do better understand, but feeling like maybe this isn't needed?

Reply:
👆 this works for an individual marker's popup. How would you allow a ClusterMarker's popup have access to the data included in its cluster?

Reply:
I thought we could do:

```
<MarkerCluster
   popup={cluster => <MyPopup />}
/>
```

where `cluster` included the data about what's included in its cluster? Is that not the case?

Reply:
Nevermind the above, I understand now.

Reply:
I like use of `data` as the prop name.